keps/sig-multicluster/4322-cluster-inventory/README.md

```diff
@@ -0,0 +1,993 @@
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+We try to summarize the most common properties that should be able to
+represent a cluster and support the above use case at the minimum scope.
+
+The target consumers of the API are users who manage the clusters and
+other tools to understand the clusters concept from the API for
+multicluster scheduling, workload distribution and cluster management.
+The API aims to provide the information for the consumers to answer the
+below questions:
+
+* Is the cluster under management?
+* How can I select a cluster?
+* Is the cluster healthy?
+* Does the cluster have certain capabilities or properties?
+* Does the cluster have sufficient resources?
+
+### Cluster Name
+
+It is required that cluster name is unique for each cluster, and it
+should also be unique among different providers (cluster manager). There
+are two option to ensure it is uniqueness
+
+#### Option 1
+
+The `metadata.name` of the cluster should never be set upon creation.
+Only `metadata.generateName` can be set.
+
+#### Option 2
+
+The `metadata.name` and `metadata.generateName` must have prefix which
+should be the same as the `spec.clusterManager.name`. Different cluster
+manager must set a different value of `spec.clusterManager.name` when
+the cluster is created.
+
+These two options can be achieved by a validating webhook or a
+`ValidatingAdmissionPolicy` resource.
+
+### Spec
+
+#### Display name
+
+It is a human-readable name of the cluster set by the consumer of the
+cluster.
+
+#### Cluster Manager
+
+An immutable field set by a cluster manager when this cluster resources
+is created by it. Each cluster manager instance should set a different
+values to this field.
+
+In addition, a predefined label with key "x-k8s.io/cluster-manager" needs
+to be added by the cluster manager upon creation. The value of the label
+MUST be the same as the name of the cluster manager. The purpose of this
+label is to make filter clusters from different cluster managers eaiser.
+
+### Status
+
+#### Version
+
+Kubernetes version of the cluster
+
+Versions of the kubernetes can let consumers understand the capability
+of the kubernetes, such as what API is supported.
+
+#### Resources
+
+Resources include capacity and allocatable, which is the sum of the
+capacity/allocatable of nodes on the cluster. It is useful for users to
+know the sizing of the cluster, and also schedulers to sort or bin pack
+the workload to clusters.
+
+#### Properties
+
+Name/value pairs to represent properties of the clusters. It could be a
+collection of ClusterProperty resources, but could also be info based on
+other implementations.
```

Comment:
the reason is trying to align with `ClusterProperty` API https://github.com/kubernetes-sigs/about-api. It seems to me it is a different use case to introduce topology information of the cluster.

Reply:
Is there a reason of why trying to align with that API? Also that was not stated in the description. So maybe more information is needed to convey your idea for the `properties` field.

Reply:
The purpose is to provide cluster level properties. Checked the https://github.com/kubernetes-sigs/node-feature-discovery/issues/1423 and `features` is also a key-value pair map. So they're aligned? 

```
// Features is the set of cluster wide features that are not reported at the node level.
// +optional
Features map[string]string `json:"features"`
```

Reply:
Yup, I think NFD https://github.com/kubernetes-sigs/node-feature-discovery/issues/1423 aligns perfectly with this KEP, a few changes, and we could merge both ideas. 
NFD has a big user base requesting for this API at the NFD level, but having it here will make a bigger impact

Reply:
Also I checked, and an average NFD `NodeFeature` CRD weights around 80k, meaning we could safely host a `ClusterInventory` of 20k nodes without reaching etcd limits of 1.5M
cc @qiujian16

Reply:
Name-value pairs match well with the ClusterClaim API and the ClusterProperty from discussion with sig-arch a couple years back.

Reply:
I agree, I am just saying we could have something like
```go
// ClusterFeatureSpec describes a ClusterFeature object.
type ClusterFeatureSpec struct {
	// Nodes is the full "raw" features data per node that has been discovered.
	// +optional
	Nodes []NodeFeatureSpec `json:"nodeFeatureSpec"` <--- from k8s-sig/nfd
	// Features is the set of cluster wide features that are not reported at the node level.
	// +optional
	Features map[string]string `json:"features"`
}
```

Reply:
@ArangoGutierrez  I wonder why we need to have a list of the node features?

Reply:
Is an Inventory of the Cluster, for HPC/AI/ML use cases, is necessary to know specific node features to be able to perform a good scheduling decision

Reply:
> is necessary to know specific node features to be able to perform a good scheduling decision

Then this comes to a question, whether we want to design a single-layer scheduling flow or a normal two-layer scheduling flow.

- single-layer scheduling: The multi-cluster scheduler (running in multi-cluster management control plane) is responsible to perform an **End-to-End** scheduling decision, i.e., selecting the perfect nodes in child/member clusters. In this way, it is true and necessary that all the specific node features in every cluster should be reported. But, please also be aware of the scheduling efficiency, heartbeat burdens, lots of data, scheduling failure, etc.

- two-layer scheduling: In this way, the multi-cluster scheduler is only responsible to select the perfect clusters, instead of nodes. We only need to make sure the target cluster is capable to run the workloads, regardless of detailed/specific node features. Normally what the multi-cluster scheduler really needs is a raw estimation/prediction to prove that cluster has enough resources, node features. With cluster labels indicating specific node features are helpful as well.

IMO, for multi-cluster scheduling, having a per-node data is superfluous. And it is too heavy when we have clusters and nodes in a large scale. I agree with @deads2k that it is better to have some well-known node features or we tweak with some kind of node feature aggregations.

Reply:
lgtm